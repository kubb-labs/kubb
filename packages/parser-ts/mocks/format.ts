import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string): Promise<string> {
  if (!source) {
    return Promise.resolve('')
  }
  return prettierFormat(source, formatOptions)
}
