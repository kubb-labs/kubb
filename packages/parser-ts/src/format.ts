import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}

/**
 * Format the generated code based on Prettier
 */
export async function format(source: string) {
  // do some basic linting with the ts compiler
  if (!source) {
    return ''
  }

  return prettierFormat(source, formatOptions)
}
