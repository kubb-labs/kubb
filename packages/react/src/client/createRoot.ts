import { createNode, ReactTemplate } from '../shared/index.ts'

import type { Logger } from '@kubb/core/utils'
import type { AppContextProps } from '../components/AppContext.tsx'
import type { DOMElement } from '../types.ts'
import type { RootType } from './types.ts'

const instances = new Map<string, ReactTemplate>()

type Props = {
  container?: DOMElement
  logger?: Logger
}

export function createRoot<Context extends AppContextProps = AppContextProps>({ container, logger }: Props = {}): RootType<Context> {
  if (!container) {
    container = createNode('kubb-root')
  }

  const instance = new ReactTemplate<Context>(container, { logger })
  instances.set(instance.id, instance)

  return {
    render(children, context?: Context) {
      return instance.render(children, context)
    },
    unmount() {
      instance.unmount()
      instances.delete(instance.id)
    },
    get output() {
      return instance.output
    },
    get file() {
      return instance.file
    },
    get files() {
      return instance.files
    },
    getFile(id: string) {
      return instance.files.find(file => file.id === id)
    },
  }
}
