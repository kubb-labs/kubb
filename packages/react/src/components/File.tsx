import { createContext } from 'react'

import { Export } from './Export.tsx'
import { Import } from './Import.tsx'

import type { KubbFile } from '@kubb/core'
import type { KubbNode } from '../types.ts'

export type FileContextProps<TMeta extends KubbFile.FileMetaBase = KubbFile.FileMetaBase> = {
  /**
   * Name to be used to dynamicly create the baseName(based on input.path).
   * Based on UNIX basename
   * @link https://nodejs.org/api/path.html#pathbasenamepath-suffix
   */
  baseName: KubbFile.BaseName
  /**
   * Path will be full qualified path to a specified file.
   */
  path: KubbFile.Path
  meta?: TMeta
}
const FileContext = createContext<FileContextProps>({} as FileContextProps)

type BasePropsWithBaseName = {
  /**
   * Name to be used to dynamicly create the baseName(based on input.path).
   * Based on UNIX basename
   * @link https://nodejs.org/api/path.html#pathbasenamepath-suffix
   */
  baseName: KubbFile.BaseName
  /**
   * Path will be full qualified path to a specified file.
   */
  path: KubbFile.Path
}

type BasePropsWithoutBaseName = {
  baseName?: never
  /**
   * Path will be full qualified path to a specified file.
   */
  path?: KubbFile.Path
}

type BaseProps = BasePropsWithBaseName | BasePropsWithoutBaseName

type Props<TMeta extends KubbFile.FileMetaBase = KubbFile.FileMetaBase> = BaseProps & {
  /**
   * Unique identifier to reuse later.
   * @default crypto.randomUUID()
   */
  id?: KubbFile.File['id']
  /**
   * This will override `process.env[key]` inside the `source`, see `getFileSource`.
   */
  env?: KubbFile.File['env']
  /**
   * This will call fileManager.add instead of fileManager.addOrAppend, adding the source when the files already exists.
   * This will also ignore the combinefiles utils
   * @default `false`
   */
  override?: KubbFile.File['override']
  meta?: TMeta
  children?: KubbNode
}

export function File<TMeta extends KubbFile.FileMetaBase = KubbFile.FileMetaBase>({ children, ...rest }: Props<TMeta>): KubbNode {
  if (!rest.baseName || !rest.path) {
    return children
  }

  return (
    <kubb-file {...rest}>
      <FileContext.Provider value={{ baseName: rest.baseName, path: rest.path, meta: rest.meta }}>{children}</FileContext.Provider>
    </kubb-file>
  )
}

type FileSourceUnionProps =
  | {
      /**
       * When path is set it will copy-paste that file as a string inside the component.
       * Children will then be ignored
       */
      path?: string
      children?: never
    }
  | {
      /**
       * When path is set it will copy-paste that file as a string inside the component.
       * Children will then be ignored
       */
      path?: never
      children?: KubbNode
    }

type FileSourceProps = FileSourceUnionProps & {
  /**
   * When true, it will return the generated import.
   * When false, it will add the import to a KubbFile instance(see fileManager).
   */
  print?: boolean
}

function FileSource({ path, print, children }: FileSourceProps): KubbNode {
  return (
    <kubb-source path={path} print={print}>
      {children}
    </kubb-source>
  )
}

File.Export = Export
File.Import = Import
File.Source = FileSource
File.Context = FileContext
