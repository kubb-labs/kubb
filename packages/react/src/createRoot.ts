import { ReactTemplate, type ReactTemplateOptions } from './ReactTemplate.tsx'

import type { RootContextProps } from './components/Root.tsx'

export function createRoot<Context extends RootContextProps = RootContextProps>(options: ReactTemplateOptions = { debug: false }): ReactTemplate<Context> {
  return new ReactTemplate<Context>(options)
}
