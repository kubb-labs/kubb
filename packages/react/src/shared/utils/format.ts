import { print } from '@kubb/parser'

import type { Language } from '../../components/Editor.tsx'
import type { DOMElement } from '../../types.ts'

export function format(text: string, node: DOMElement): string {
  const attributes = node.attributes as React.ComponentProps<(typeof Language)>
  try {
    if (attributes.value === 'typescript') {
      return print([], { source: text, noEmitHelpers: false })
    }

    return text
  } catch (e) {
    console.log(e)
  }

  return text
}
