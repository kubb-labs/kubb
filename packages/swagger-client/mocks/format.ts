import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string) {
  if (!source) {
    return ''
  }
  return prettierFormat(source, formatOptions)
}
