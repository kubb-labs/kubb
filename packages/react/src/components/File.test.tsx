import path from 'node:path'

import { format } from '../../mocks/format.ts'
import { createRoot } from '../client/createRoot.ts'
import { File } from './File.tsx'

describe('<File/>', () => {
  const mocksPath = path.resolve(__dirname, '../../mocks')

  test('render File', () => {
    const Component = () => {
      return <File baseName="test.ts" path="path" />
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toStrictEqual('')
  })

  test('render File with Import and Export', () => {
    const Component = () => {
      return (
        <File baseName="test.ts" path="path">
          <File.Import name={'React'} path="react" />
          <File.Export asAlias path="./index.ts" />
        </File>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.files.at(0)?.exports).toStrictEqual([
      {
        asAlias: true,
        isTypeOnly: false,
        name: undefined,
        path: './index.ts',
      },
    ])

    expect(root.files.at(0)?.imports).toStrictEqual([
      {
        isTypeOnly: false,
        name: 'React',
        path: 'react',
        root: undefined,
      },
    ])
  })

  test('render File with Import and Export and print imports/exports', async () => {
    const Component = () => {
      return (
        <File baseName="test.ts" path="path">
          <File.Import name={'React'} path="react" print />
          <br />
          <File.Export asAlias path="./index.ts" print />
        </File>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(await format(root.output)).toMatchSnapshot()
  })

  test('render File with source', () => {
    const Component = () => {
      return (
        <File baseName="test.ts" path="path">
          <File.Source>test</File.Source>
        </File>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toStrictEqual('test')
  })
  test('render File with source and path, print', async () => {
    const Component = () => {
      return (
        <File baseName="test.ts" path="path">
          <File.Source path={path.resolve(mocksPath, './test.ts')} print />
        </File>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(await format(root.output)).toMatchSnapshot()
  })

  test('render File with multiple sources', async () => {
    const Component = () => {
      return (
        <File baseName="test.ts" path="path">
          <File.Source path={path.resolve(mocksPath, './test.ts')} print />
          <File.Source print>
            {`
            const test = 2;
            `}
          </File.Source>
        </File>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(await format(root.output)).toMatchSnapshot()
  })

  test('render multiple Files', async () => {
    const Component = () => {
      return (
        <>
          <File baseName="test.ts" path="./">
            <File.Source print>
              {`
            const test = 1;
            `}
              <File.Import name="node" path="node" />
            </File.Source>
          </File>
          <File baseName="test2.ts" path="./">
            <File.Source print>
              {`
            const test2 = 2;
            `}
            </File.Source>
          </File>
        </>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(await format(root.output)).toMatchSnapshot()

    expect(root.files.length).toBe(2)

    expect(await format(root.files[0]?.source)).toMatchSnapshot()

    expect(root.files[0]?.imports).toStrictEqual([
      {
        isTypeOnly: false,
        name: 'node',
        path: 'node',
        root: undefined,
      },
    ])

    expect(await format(root.files[1]?.source)).toMatchSnapshot()
  })

  test('render File with multiple sources', async () => {
    const Component = () => {
      return (
        <File baseName="test.ts" path="path">
          <File.Source path={path.resolve(mocksPath, './test.ts')} print />
          <File.Source print>
            {`
            const test = 2;
            `}
          </File.Source>
        </File>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(await format(root.output)).toMatchSnapshot()
  })

  test('render multiple Files', async () => {
    const Component = () => {
      return (
        <>
          <File baseName="test.ts" path="./">
            <File.Source print>
              {`
            const test = 1;
            `}
              <File.Import name="node" path="node" />
            </File.Source>
          </File>
          <File baseName="test2.ts" path="./">
            <File.Source print>
              {`
            const test2 = 2;
            `}
            </File.Source>
          </File>
        </>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(await format(root.output)).toMatchSnapshot()

    expect(root.files.length).toBe(2)

    expect(await format(root.files[0]?.source)).toMatchSnapshot()

    expect(root.files[0]?.imports).toStrictEqual([
      {
        isTypeOnly: false,
        name: 'node',
        path: 'node',
        root: undefined,
      },
    ])

    expect(await format(root.files[1]?.source)).toMatchSnapshot()
  })
})
