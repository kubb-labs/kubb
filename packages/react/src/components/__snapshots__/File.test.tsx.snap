// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<File/> > render File with Import and Export and print imports/exports 1`] = `
"import React from 'react'
export * from './index.ts'
"
`;

exports[`<File/> > render File with multiple sources 1`] = `
"export function test() {
  return true
}
const test = 2
"
`;

exports[`<File/> > render File with multiple sources 2`] = `
"export function test() {
  return true
}
const test = 2
"
`;

exports[`<File/> > render File with source and path, print 1`] = `
"export function test() {
  return true
}
"
`;

exports[`<File/> > render multiple Files 1`] = `
"const test = 1
const test2 = 2
"
`;

exports[`<File/> > render multiple Files 2`] = `
"const test = 1
const test2 = 2
"
`;

exports[`<File/> > render multiple Files 3`] = `
"const test = 1
"
`;

exports[`<File/> > render multiple Files 4`] = `
"const test2 = 2
"
`;

exports[`<File/> > render multiple Files 5`] = `
"const test = 1
const test2 = 2
"
`;

exports[`<File/> > render multiple Files 6`] = `
"const test = 1
const test2 = 2
"
`;

exports[`<File/> > render multiple Files 7`] = `
"const test = 1
"
`;

exports[`<File/> > render multiple Files 8`] = `
"const test2 = 2
"
`;
